packages/browser-repl/src/iframe-runtime/iframe-runtime.ts

```diff
@@ -32,24 +33,19 @@ export class IframeRuntime implements Runtime {
   }
 
   async evaluate(code: string): Promise<ShellResult> {
-    if (!this.openContextRuntime) {
-      await this.initialize();
-    }
-
-    return await this.openContextRuntime.evaluate(code);
+    const runtime = await this.initialize();
+    return await runtime.evaluate(code);
   }
 
   async getCompletions(code: string): Promise<Completion[]> {
-    if (!this.openContextRuntime) {
-      await this.initialize();
-    }
-
-    return await this.openContextRuntime.getCompletions(code);
+    const runtime = await this.initialize();
+    return await runtime.getCompletions(code);
   }
 
-  async initialize(): Promise<void> {
-    if (this.iframe) {
-      return;
+  async initialize(): Promise<OpenContextRuntime> {
+    if (this.readyPromise !== null) {
+      await this.readyPromise;
+      return this.openContextRuntime as OpenContextRuntime;
     }
```

Comment:
This might have been a bug? The previous code checks for `this.iframe`, but the presence of the property does not imply that the `onload` event that we wait for below already happened.

Reply:
I'm only impressed that you managed to spot and fix this weird concurrency issue among a mountain of code 😄 . Yep that could have been a bug, just never happened cause the first evaluation is normally happening quite late from the initialisation. 💯

Reply:
I really spent far too much time on the Worker thread implementation in Node.js not to be on the lookup for concurrency issues :see_no_evil: